apps/axistream/src/message.hh

```diff
@@ -0,0 +1,498 @@
+    }
+
+};
+
+template <>
+struct EndiannessConverter<Endianness::BIG> {
+    template <int W>
+    static ap_uint<W> convert(const ap_uint<W>& x) {
+        HLS_PRAGMA(HLS inline);
+        return flip_endianness(x);
+    }
+};
+
+/*
+ * Helper type to select between ap_uint and ap_int
+ */
+template <int W, Sign S>
+struct TypeSelector {};
+
+template <int W>
+struct TypeSelector<W, Sign::SIGNED> {
+    typedef ap_int<W> type;
+};
+
+template <int W>
+struct TypeSelector<W, Sign::UNSIGNED> {
+    typedef ap_uint<W> type;
+};
+
+template <> struct TypeSelector<8, Sign::SIGNED> { typedef int8_t type; };
+template <> struct TypeSelector<16, Sign::SIGNED> { typedef int16_t type; };
+template <> struct TypeSelector<32, Sign::SIGNED> { typedef int32_t type; };
+template <> struct TypeSelector<64, Sign::SIGNED> { typedef int64_t type; };
+template <> struct TypeSelector<8, Sign::UNSIGNED> { typedef uint8_t type; };
+template <> struct TypeSelector<16, Sign::UNSIGNED> { typedef uint16_t type; };
+template <> struct TypeSelector<32, Sign::UNSIGNED> { typedef uint32_t type; };
+template <> struct TypeSelector<64, Sign::UNSIGNED> { typedef uint64_t type; };
+
+/*
+ * Type definitions. All types must define:
+ * width: the width in bits
+ * raw: the type of a raw field - ap_uint<width>. Note that even
+ * though the raw value of a field might have the same type as the nominal
+ * type (given by type, below), it may not be equal in value.
+ * For example, the endianness may be flipped.
+ * type: the type of a field when used normally. For integer fields,
+ * this is ap_int<width> or ap_uint<width>. For enum fields,
+ * this is the enum type. For composite fields (fields with a struct
+ * or union type), this is a message of that type.
+ * to_raw: function from the nominal type to the raw type
+ * from_raw: function from the raw type to the nominal type
+ */
+
+template <int W, Sign S, Endianness E>
+struct Int {
+    static const int width = W;
+
+    typedef ap_uint<width> raw;
+    typedef typename TypeSelector<W, S>::type type;
+
+    static raw to_raw(const type& x) {
+        HLS_PRAGMA(HLS inline);
+        return EndiannessConverter<E>::convert(raw(x));
+    }
+
+    static type from_raw(const raw& x) {
+        HLS_PRAGMA(HLS inline);
+        return EndiannessConverter<E>::convert(x);
+    }
+};
+
+template <int W, Sign S> using lx = Int<W, S, Endianness::LITTLE>;
+template <int W, Sign S> using bx = Int<W, S, Endianness::BIG>;
+
+template <int W> using li = lx<W, Sign::SIGNED>;
+template <int W> using lu = lx<W, Sign::UNSIGNED>;
+template <int W> using bi = bx<W, Sign::SIGNED>;
+template <int W> using bu = bx<W, Sign::UNSIGNED>;
+
+using i8 = li<8>;
+using li16 = li<16>;
+using li32 = li<32>;
+using li64 = li<64>;
+using bi16 = bi<16>;
+using bi32 = bi<32>;
+using bi64 = bi<64>;
+
+using u8 = lu<8>;
+using lu16 = lu<16>;
+using lu32 = lu<32>;
+using lu64 = lu<64>;
+using bu16 = bu<16>;
+using bu32 = bu<32>;
+using bu64 = bu<64>;
```

Comment:
I tried to avoid defining specific descriptors like this.   Although I don't think it's super difficult to do it this way, I was hoping to have people define their header structures just using regular types.

Reply:
I tried to figure out how to do that, but it just worked better if everything was in this system.

In the original library, you just used `newfield` right? https://github.com/Xilinx/HLS_packet_processing/blob/master/apps/common/ip.hpp#L2182:
```c++
typedef newfield<ap_uint<16>, boost::mpl::string<'type'> > etherType;
```
Or was there another way?

Reply:
I tried to figure out how to do that, but it just worked better if everything was in this system.

In the original library, you just used `newfield` right? https://github.com/Xilinx/HLS_packet_processing/blob/master/apps/common/ip.hpp#L2182:
```c++
typedef newfield<ap_uint<16>, boost::mpl::string<'type'> > etherType;
```
Or was there another way?

Reply:
This is fundamentally the same as what you are doing.  Lookup is by type comparison in a sequence of types.  I agree that it's probably simpler the way you've defined these types, rather than using boost::mpl, which is a significant library dependance and also has some syntactic overhead.

Reply:
I see, so in the original library, you made a field of type `ap_uint<16>`, where now you have to do `lu16`.